From a4b8ce98d8158e11eeb999938194ff84edd9d9bf Mon Sep 17 00:00:00 2001
From: Isuru Fernando <isuruf@gmail.com>
Date: Thu, 9 Apr 2020 18:08:39 +0000
Subject: [PATCH 1/5] Remove linking liblldb to libpython

liblldb doubles as an extension module too, but python
executable is linked to static python. When the extension module
is loaded both the shared libpython and the static libpython symbols
are loaded which leads to a crash in OSX.

To fix this, dependence on PYTHON_LIBRARY is removed from liblldb
and added to its interfaces to make sure the executables like
lldb works fine.

Thanks to feedback from @saulshanabrook

Co-authored-by: Uwe L. Korn <uwelk@xhochy.com>
Co-authored-by: H. Vetinari <h.vetinari@gmx.com>
---
 lldb/cmake/modules/AddLLDB.cmake | 3 +++
 lldb/source/API/CMakeLists.txt   | 6 ++++++
 2 files changed, 9 insertions(+)

diff --git a/lldb/cmake/modules/AddLLDB.cmake b/lldb/cmake/modules/AddLLDB.cmake
index b6ad8d46380b..40988b0e121f 100644
--- a/lldb/cmake/modules/AddLLDB.cmake
+++ b/lldb/cmake/modules/AddLLDB.cmake
@@ -386,6 +386,9 @@ function(add_lldb_executable name)
   endif()
 
   list(APPEND LLVM_LINK_COMPONENTS ${ARG_LINK_COMPONENTS})
+  if (NOT MSVC)
+    list(APPEND LLVM_LINK_COMPONENTS ${Python3_LIBRARIES})
+  endif()
   add_llvm_executable(${name}
     ${pass_NO_INSTALL_RPATH}
     ${ARG_UNPARSED_ARGUMENTS}
diff --git a/lldb/source/API/CMakeLists.txt b/lldb/source/API/CMakeLists.txt
index 15c52aab4136..9fb28b42e7cb 100644
--- a/lldb/source/API/CMakeLists.txt
+++ b/lldb/source/API/CMakeLists.txt
@@ -249,6 +249,12 @@ elseif (LLDB_EXPORT_ALL_SYMBOLS OR LLDB_ENABLE_DYNAMIC_SCRIPTINTERPRETERS)
 endif()
 
 if(NOT WIN32 OR MINGW)
+  target_link_libraries(liblldb INTERFACE ${Python3_LIBRARIES})
+  if ( CMAKE_SYSTEM_NAME MATCHES "Darwin" )
+    target_link_options(liblldb PRIVATE -undefined dynamic_lookup)
+  else()
+    target_link_options(liblldb PRIVATE "-Wl,--unresolved-symbols=ignore-all")
+  endif()
   set_target_properties(liblldb
     PROPERTIES
     OUTPUT_NAME lldb
